### PR TITLE
New version of Colombe/Sendmail, monadic implementation

### PR DESCRIPTION
**Pull request by dinosaure**
The most interesting part is about how to describe an SMTP state machine.
- [x] Monadic view of the SMTP state machine
- [x] `Sendmail` implementation with the new core
- [x] STARTTLS with a specific context
- [x] Others extensions like `8BITMIME`
- [x] Tests and Fuzzers

### CODE REVIEW

**Comment by dinosaure** [comment]
Good ratio.

File: sendmail/sendmail.mli [hunk 1]
```diff
@@ -2,63 +2,35 @@ open Colombe.Sigs
...
+  { username : string
+  ; password : string
+  ; mechanism : mechanism }
+
+type ('a, 's) stream = unit -> ('a option, 's) io
+
+type error =
+  [ Request.Decoder.error
+  | Reply.Decoder.error
+  | `Unexpected_response of (int * string list)
```
Comment (avsm) [inline sendmail/sendmail.mli]: unnecessary boxing here 
```suggestion
  | `Unexpected_response of int * string list
```

File: src/decoder.ml [hunk 2]
```diff
@@ -19,41 +19,51 @@ let decoder_from x =
...
-
-let safe k decoder : 'v state =
-  try k decoder with Leave info -> Error info
+  | `End_of_input -> Fmt.string ppf "End of input"
+  | `Expected_char chr -> Fmt.pf ppf "Expected char: %02x" (Char.code chr)
+  | `Unexpected_char chr -> Fmt.pf ppf "Unexpected char: %02x" (Char.code chr)
+  | `Expected_string s -> Fmt.pf ppf "Expected_string: %s" s
+  | `Expected_eol -> Fmt.string ppf "Expected end-of-line"
+  | `Expected_eol_or_space -> Fmt.string ppf "Expected end-of-line or space"
+  | `No_enough_space -> Fmt.string ppf "No enough space"
```
Comment (avsm) [inline src/decoder.ml]:
```suggestion
  | `No_enough_space -> Fmt.string ppf "Not enough space"
```

File: src/decoder.ml [hunk 3]
```diff
@@ -19,41 +19,51 @@ let decoder_from x =
...
-  | Assert_predicate of (char -> bool)
-  | Invalid_code of int
+  [ `End_of_input
+  | `Expected_char of char
+  | `Unexpected_char of char
+  | `Expected_string of string
+  (* | Invalid_command of string *)
+  | `Expected_eol
+  | `Expected_eol_or_space
+  | `No_enough_space
```
Comment (avsm) [inline src/decoder.ml]: unrelated to this PR, but should be "not_enough_space"

**Comment by dinosaure** [comment]
Ok final global version is done. We need to polish some returned code from the process described by `sendmail` and `sendmail_with_tls` but the global view is much more better than before - another good news is the ability to wrap the mail into a scheduler's stream (like `Lwt_stream` or `Async.Pipe.t`).

**Comment by dinosaure** [comment]
Ok, just need to integrate changes from @avsm and ready to merge!

**Pull request merged**